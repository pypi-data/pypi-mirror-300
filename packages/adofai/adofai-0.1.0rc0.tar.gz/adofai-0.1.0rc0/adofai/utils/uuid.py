# coding=utf-8
"""UUID 工具包，主要包含玩家 UUID 相关的工具"""
__all__ = ["uuid_to_str", "offline_uuid", "pcl_offline_uuid"]

from hashlib import md5
from uuid import UUID

from adofai import GameId, GameName


def uuid_to_str(src: UUID | str, signed: bool = False) -> str:
    """
    从 UUID 对象或字符串生成小写的 UUID

    返回值与 str(self) 或 self.hex 相同

    >>> uuid_to_str("B50AD385829D3141A2167E7D7539BA7F", signed=True)
    'b50ad385-829d-3141-a216-7e7d7539ba7f'

    :param src: 原始的 UUID 对象或 UUID 格式文本
    :param signed: 布尔值，返回的文本是否有符号，默认为无符号
    :return: 小写的 UUID 文本
    """
    if isinstance(src, str):
        src = UUID(src)
    return str(src) if signed else src.hex


def offline_uuid(name: GameName) -> GameId:
    """
    从离线玩家的用户名生成原版规则的离线 UUID

    >>> offline_uuid(GameName("Notch"))
    UUID('b50ad385-829d-3141-a216-7e7d7539ba7f')
    >>> offline_uuid(GameName("Notch")).hex
    'b50ad385829d3141a2167e7d7539ba7f'

    :param name: 离线玩家的用户名
    :return: GameId 对象，表示专业服务器离线玩家的 UUID
    """
    return GameId(UUID(md5(("OfflinePlayer:" + name).encode()).hexdigest(), version=3))


def pcl_hash(src: str) -> int:
    """PCL2 使用的哈希模型
    :param src: 待计算哈希的字符串
    :return: 计算的哈希结果
    """
    num1 = 5381
    for i in src:
        num1 = (num1 << 5) ^ num1 ^ ord(i)
    ans = f"{num1 ^ 12218072394304324399:X}"
    return int.from_bytes(bytes.fromhex(ans[len(ans) - 16:]))


def pcl_offline_uuid(name: GameName) -> GameId:
    """PCL2 使用的离线 UUID 生成模型
    :param name: 离线玩家的用户名
    :return: GameId 对象，表示 PCL2 启动器离线房主玩家的 UUID
    """
    pad = lambda src, char, length: src[:length].rjust(length, char)
    full_uuid = pad(f"{len(name):X}", "0", 16) + pad(f"{pcl_hash(name):X}", "0", 16)
    full_uuid = full_uuid[0:12] + "3" + full_uuid[13:16] + "9" + full_uuid[17:32]
    return GameId(UUID(full_uuid))
