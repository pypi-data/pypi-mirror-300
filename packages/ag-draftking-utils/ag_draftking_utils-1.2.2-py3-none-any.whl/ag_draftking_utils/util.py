import datetime 
import time
import os
import inspect
import glob 
import pandas as pd


def get_current_chicago_time():
    """
    Returns datetime object of the current chicago time, needs to be adjusted during DST
    """
    return datetime.datetime.utcnow() - datetime.timedelta(hours=5)

def run_query(query, conn, has_game_date=True):
    start = time.time()
    df = pd.read_sql_query(query, conn)
    end = time.time()
    n = end - start 
    rows, cols = df.shape
    if has_game_date:
        df['GAME_DATE'] = pd.to_datetime(df['GAME_DATE'])
    print(f'Returned {rows} rows, {cols} cols, took {n} seconds.')
    return df

def run_query_from_file(file, conn, has_game_date=True):
    """
    Input:
        file: str - name of file containing SQL query
        conn: SQLAlchemy object - database connection object
    Output:
        df: pandas Dataframe containing the queries output
    """
    s = time.time()
    with open(file, 'r') as f:
        q = f.read().strip()
    df = pd.read_sql_query(q, conn)
    if has_game_date:
        df['GAME_DATE'] = pd.to_datetime(df['GAME_DATE'])
    e = time.time()
    t = '{:.1f}'.format(e-s)

    print(f'Successfully Queried Data from {file}, took {t} seconds.')
    return df 


def write_to_db(df, table_name, conn):
    start = time.time()
    df.to_sql(table_name, conn, if_exists='append', index=False)
    end = time.time()
    rows, cols = df.shape
    n = end - start 
    print(f'Saved {rows} rows, {cols} cols to table {table_name}, took {n} seconds')


def get_todays_df(df, today_column='GAME_DATE'):
    return df[df[today_column].dt.date == datetime.date.today()]


def time_function(func):
    """
    Decorator that reports the execution time of a function
    """
    def wrap(*args, **kwargs):
        start = time.time()
        result = func(*args, **kwargs)
        end = time.time()
        t = '{:.1f}'.format(end-start)

        print(f'Function {func.__name__} took {t} seconds')
        return result
    return wrap


def time_function_simple(func):
    """
    Decorator that reports the execution time of a function
    """
    def wrap(*args, **kwargs):
        start = time.time()
        result = func(*args, **kwargs)
        end = time.time()
        t = '{:.1f}'.format(end-start)

        
        bound_args = inspect.signature(func).bind(*args, **kwargs)
        bound_args.apply_defaults()
        target_args = dict(bound_args.arguments)
        
        kv = ', '.join([f'{x[0]}={x[1]}' for x in list(zip(target_args.keys(), target_args.values()))])

        print(f'Function {func.__name__} took {t} seconds, ran with arguments {kv}')
        return result
    return wrap

def get_most_recent_file_from_directory(directory):
    """
    Input:
        directory: str - (i.e. '/Users/t2/Desktop')
    Output:
        str - fully-qualified filename, i.e. /Users/t2/Desktop/fantasy/GitDK/DraftKings/data/aggregated_data/2022-01-03_13:06:14.319307.csv
    """
    list_of_files = glob.glob(f'{directory}/*')
    return max(list_of_files, key=os.path.getctime)

def write_pandas_csv(df, file, index=False):
    rows, cols = df.shape
    start = time.time()
    df.to_csv(file, index=index)
    end = time.time()
    print(f"Saved {n} rows, {cols} cols to {file}, took {end-start} seconds")

def read_pandas_csv(file):
    start = time.time()
    df = pd.read_csv(file)
    end = time.time()
    rows, cols = df.shape 
    print(f"Read in {rows} rows and {cols} columns from file {file}, took {end-start} seconds")
    return df 

def find_latest_run(p):
    """
    Look in a log directory and return the latest file. all files must be in format:
         "%Y-%m-%d_%H:%M:%S.log"
    :param p: str - Path
    :return: most recent datetime object. If there are no files in the directory,
        this will default to January 4, 2005.
    """
    dates = []
    for file in os.listdir(p):
        with open(os.path.join(p, file), 'r') as f:
            x = f.read()
        # ignore empty file
        if len(x) == 0:
            continue

        date_info = file.split('.')[0]
        date = datetime.datetime.strptime(date_info, "%Y-%m-%d_%H:%M:%S")
        dates.append(date)

    if len(dates) == 0:
        # Default to arbitrary date, January 4, 2005
        return datetime.datetime(2005, 1, 4)

    most_recent = sorted(dates, reverse=True)[0]
    return most_recent