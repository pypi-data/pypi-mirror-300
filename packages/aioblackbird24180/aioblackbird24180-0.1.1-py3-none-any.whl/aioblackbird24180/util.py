"""Utilities for aioblackbird24180."""

import json


def parse_response(response: str) -> dict:  # type: ignore[type-arg]
    """
    Parse the response.

    Parameters:
        `response`: the response to parse as string.
    """
    # Remove the leading and trailing parentheses
    response = response.strip().strip("()")
    # The status response uses single quotes, but json requires double quotes
    response = response.replace("'", '"')
    return json.loads(response)  # type: ignore[no-any-return]
