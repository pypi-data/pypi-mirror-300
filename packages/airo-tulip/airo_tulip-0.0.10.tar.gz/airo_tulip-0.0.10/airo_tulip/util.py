import math


def clip(value: float, maximum: float, minimum: float) -> float:
    return min(max(value, minimum), maximum)


def clip_angle(angle: float) -> float:
    if angle < -math.pi:
        return angle + (2 * math.pi)
    elif angle > math.pi:
        return angle - (2 * math.pi)
    else:
        return angle

def get_shortest_angle(angle1: float, angle2: float) -> float:
    return math.atan2(math.sin(angle1 - angle2), math.cos(angle1 - angle2))

def sign(a: float) -> float:
    return 1.0 if a >= 0 else -1.0
