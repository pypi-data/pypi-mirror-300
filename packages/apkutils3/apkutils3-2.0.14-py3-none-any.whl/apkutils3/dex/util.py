# Copyright 2015 Google Inc. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# Modified by:
#     - @Young-Lord <ly-niko@qq.com>


def keysToRanges(d: dict, limit: int) -> dict:
    starts = sorted(d)
    for s, e in zip(starts, starts[1:] + [limit]):
        for k in range(s, e):
            d[k] = d[s]
    return d


def signExtend(val: int, size: int) -> int:
    if val & (1 << (size - 1)):
        val -= 1 << size
    return val


def s16(val: int) -> int:
    val %= 1 << 16
    if val >= 1 << 15:
        val -= 1 << 16
    return val


def s32(val: int) -> int:
    val %= 1 << 32
    if val >= 1 << 31:
        val -= 1 << 32
    return val


def s64(val: int) -> int:
    val %= 1 << 64
    if val >= 1 << 63:
        val -= 1 << 64
    return val
