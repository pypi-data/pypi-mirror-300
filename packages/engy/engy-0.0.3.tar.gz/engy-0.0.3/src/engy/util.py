import os
import pickle
from pathlib import Path
from typing import List

from .llm import ChatMessage


def assert_file_exists_and_read(filename):
    # Assert that the file exists
    assert os.path.exists(filename), f"File '{filename}' does not exist"

    # Read the file content
    try:
        with open(filename, 'r') as file:
            content = file.read()
        return content
    except Exception as e:
        assert False, f"Failed to read file '{filename}': {str(e)}"


def save_history(history, prefix='default'):
    history_filename = Path(f'{prefix}.history.pickle')
    with open(history_filename, 'wb') as f:
        return pickle.dump(history, f)


def load_history(prefix='default') -> List[ChatMessage]:
    history_filename = Path(f'{prefix}.history.pickle')
    if history_filename.exists():
        return pickle.loads(history_filename.read_bytes())
    return []


_PRODUCE_FILE_MAPPING = {
    'AGENTS_DESIGN': 'agents_design.json',
    'TASKS_DESIGN': 'tasks_design.json',
    'SERVER_PYTHON_CODE': 'server.py',
    'INDEX_HTML_CODE': 'index.html',
    'RUN_BASH_CODE': 'run.sh',
    'BACKEND_DESIGN': 'backend_design.txt',
    'FRONTEND_DESIGN': 'frontend_design.txt',
    'AGENTS_PYTHON_CODE': 'agents.py',
}


def parse_block(block_name, response_text):
    if f'<{block_name}>' in response_text:
        try:
            block_content = response_text.split(f'<{block_name}>')[
                1].split(f'</{block_name}>')[0]
            return block_content
        except:
            pass
    return None


def produce_files(response_text):
    for block_name, filename in _PRODUCE_FILE_MAPPING.items():
        block_content = parse_block(block_name, response_text)
        if block_content is not None:
            with open(filename, 'w') as f:
                f.write(block_content)
