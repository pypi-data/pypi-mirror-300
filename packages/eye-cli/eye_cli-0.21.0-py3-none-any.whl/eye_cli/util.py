import subprocess

import typer
from rich import print
from rich.padding import Padding


def abort(s):
    header = color("Exiting... ", "red")
    message(header + s, padding="around")
    raise typer.Abort()


def message(the_message, padding=None, indent: int = 0):
    if indent:
        the_message = Padding.indent(the_message, indent)

    if padding == "around":
        the_message = Padding(the_message, pad=(1, 2, 1, 2))
    elif padding == "above":
        the_message = Padding(the_message, pad=(1, 2, 0, 2))
    elif padding == "below":
        the_message = Padding(the_message, pad=(0, 2, 1, 2))

    print(the_message)


def color(s, color):
    return f"[bold {color}]{s}[/]"


def cmd(args):
    return subprocess.run(args, check=True, capture_output=True, encoding="utf-8")
