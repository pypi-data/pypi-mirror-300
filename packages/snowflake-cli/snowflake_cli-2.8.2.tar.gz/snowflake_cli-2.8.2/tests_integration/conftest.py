# Copyright (c) 2024 Snowflake Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from __future__ import annotations

import functools
import json
import shutil
import tempfile
from contextlib import contextmanager
from dataclasses import dataclass
from json import JSONDecodeError
from pathlib import Path
from typing import Any, Dict, List, Optional

import pytest
import yaml

from snowflake.cli.api.cli_global_context import cli_context_manager
from snowflake.cli.app.cli_app import app_factory
from typer import Typer
from typer.testing import CliRunner

from tests.conftest import clean_logging_handlers_fixture  # noqa: F401
from tests.testing_utils.fixtures import (
    alter_snowflake_yml,  # noqa: F401
    snowflake_home,
)
from tests.testing_utils.files_and_dirs import merge_left

pytest_plugins = [
    "tests_common",
    "tests_integration.testing_utils",
    "tests_integration.snowflake_connector",
]

TEST_DIR = Path(__file__).parent
DEFAULT_TEST_CONFIG = "connection_configs.toml"


@dataclass
class CommandResult:
    exit_code: int
    json: Optional[List[Dict[str, Any]] | Dict[str, Any]] = None
    output: Optional[str] = None
    stderr: Optional[str] = None


class TestConfigProvider:
    def __init__(self, temp_dir_with_configs: Path):
        self._temp_dir_with_configs = temp_dir_with_configs

    def get_config_path(self, file_name: str) -> Path:
        return self._temp_dir_with_configs / file_name


@pytest.fixture(scope="session")
def test_snowcli_config_provider():
    with tempfile.TemporaryDirectory() as td:
        temp_dst = Path(td) / "config"
        shutil.copytree(TEST_DIR / "config", temp_dst)
        yield TestConfigProvider(temp_dst)


@pytest.fixture(scope="session")
def test_root_path():
    return TEST_DIR


class SnowCLIRunner(CliRunner):
    def __init__(self, app: Typer, test_config_provider: TestConfigProvider):
        super().__init__()
        self.app = app
        self._test_config_provider = test_config_provider
        self._test_config_path = self._test_config_provider.get_config_path(
            DEFAULT_TEST_CONFIG
        )

    def use_config(self, config_file_name: str) -> None:
        self._test_config_path = self._test_config_provider.get_config_path(
            config_file_name
        )

    @functools.wraps(CliRunner.invoke)
    def invoke(self, *a, **kw):
        if "catch_exceptions" not in kw:
            kw.update(catch_exceptions=False)
        return super().invoke(self.app, *a, **kw)

    def invoke_with_config(self, args, **kwargs) -> CommandResult:
        result = self.invoke(
            [
                "--config-file",
                self._test_config_path,
                *args,
            ],
            **kwargs,
        )

        if result.output == "" or result.output.strip() == "Done":
            return CommandResult(result.exit_code, json=[])
        try:
            return CommandResult(
                result.exit_code, json.loads(result.output), output=result.output
            )
        except JSONDecodeError:
            return CommandResult(result.exit_code, output=result.output)

    def invoke_json(self, args, **kwargs) -> CommandResult:
        return self.invoke_with_config([*args, "--format", "JSON"], **kwargs)

    def invoke_with_connection_json(
        self, args, connection: str = "integration", **kwargs
    ) -> CommandResult:
        return self.invoke_json([*args, "-c", connection], **kwargs)

    def invoke_with_connection(
        self, args, connection: str = "integration", **kwargs
    ) -> CommandResult:
        return self.invoke_with_config([*args, "-c", connection], **kwargs)


@pytest.fixture
def runner(test_snowcli_config_provider):
    app = app_factory()
    yield SnowCLIRunner(app, test_snowcli_config_provider)


class QueryResultJsonEncoderError(RuntimeError):
    def __init__(self, output: str):
        super().__init__(f"Can not parse query result:\n{output}")


@pytest.fixture
def project_directory(temporary_working_directory, test_root_path):
    @contextmanager
    def _temporary_project_directory(
        project_name,
        merge_project_definition: Optional[dict] = None,
        subpath: Optional[Path] = None,
    ):
        test_data_file = test_root_path / "test_data" / "projects" / project_name
        project_dir = temporary_working_directory
        if subpath:
            project_dir = temporary_working_directory / subpath
            project_dir.mkdir(parents=True)
        shutil.copytree(test_data_file, project_dir, dirs_exist_ok=True)
        if merge_project_definition:
            with Path("snowflake.yml").open("r") as fh:
                project_definition = yaml.safe_load(fh)
            merge_left(project_definition, merge_project_definition)
            with open(Path(project_dir) / "snowflake.yml", "w") as file:
                yaml.dump(project_definition, file)

        yield project_dir

    return _temporary_project_directory


@pytest.fixture(autouse=True)
def reset_global_context_after_each_test(request):
    cli_context_manager.reset()
    yield


# This automatically used fixture isolates default location
# of config files from user's system.
@pytest.fixture(autouse=True)
def isolate_snowflake_home(snowflake_home):
    yield snowflake_home
