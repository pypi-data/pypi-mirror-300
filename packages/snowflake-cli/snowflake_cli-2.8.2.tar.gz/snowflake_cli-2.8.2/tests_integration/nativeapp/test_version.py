# Copyright (c) 2024 Snowflake Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import uuid

from snowflake.cli.api.project.util import generate_user_env

from tests.project.fixtures import *
from tests_integration.test_utils import (
    pushd,
    contains_row_with,
    not_contains_row_with,
    row_from_snowflake_session,
)
from tests_integration.test_utils import (
    enable_definition_v2_feature_flag,
)

USER_NAME = f"user_{uuid.uuid4().hex}"
TEST_ENV = generate_user_env(USER_NAME)


# Tests a simple flow of an existing project, executing snow app version create, drop and teardown, all with distribution=internal
@pytest.mark.integration
@enable_definition_v2_feature_flag
@pytest.mark.parametrize(
    "project_definition_files", ["integration", "integration_v2"], indirect=True
)
def test_nativeapp_version_create_and_drop(
    runner,
    snowflake_session,
    project_definition_files: List[Path],
):
    project_name = "integration"
    project_dir = project_definition_files[0].parent
    with pushd(project_dir):
        result_create = runner.invoke_with_connection_json(
            ["app", "version", "create", "v1", "--force", "--skip-git-check"],
            env=TEST_ENV,
        )
        assert result_create.exit_code == 0

        try:
            # package exist
            package_name = f"{project_name}_pkg_{USER_NAME}".upper()
            assert contains_row_with(
                row_from_snowflake_session(
                    snowflake_session.execute_string(
                        f"show application packages like '{package_name}'",
                    )
                ),
                dict(name=package_name),
            )

            # app package contains version v1
            expect = snowflake_session.execute_string(
                f"show versions in application package {package_name}"
            )
            actual = runner.invoke_with_connection_json(
                ["app", "version", "list"], env=TEST_ENV
            )
            assert actual.json == row_from_snowflake_session(expect)

            result_drop = runner.invoke_with_connection_json(
                ["app", "version", "drop", "v1", "--force"],
                env=TEST_ENV,
            )
            assert result_drop.exit_code == 0
            actual = runner.invoke_with_connection_json(
                ["app", "version", "list"], env=TEST_ENV
            )
            assert len(actual.json) == 0

            # make sure we always delete the package
            result = runner.invoke_with_connection_json(
                ["app", "teardown"],
                env=TEST_ENV,
            )
            assert result.exit_code == 0

            expect = snowflake_session.execute_string(
                f"show application packages like '{package_name}'"
            )
            assert not_contains_row_with(
                row_from_snowflake_session(expect), {"name": package_name}
            )

        finally:
            # teardown is idempotent, so we can execute it again with no ill effects
            result = runner.invoke_with_connection_json(
                ["app", "teardown", "--force"],
                env=TEST_ENV,
            )
            assert result.exit_code == 0


# Tests upgrading an app from an existing loose files installation to versioned installation.
@pytest.mark.integration
@enable_definition_v2_feature_flag
@pytest.mark.parametrize(
    "project_definition_files", ["integration", "integration_v2"], indirect=True
)
def test_nativeapp_upgrade(
    runner,
    snowflake_session,
    project_definition_files: List[Path],
):
    project_name = "integration"
    project_dir = project_definition_files[0].parent
    with pushd(project_dir):
        runner.invoke_with_connection_json(
            ["app", "run"],
            env=TEST_ENV,
        )
        runner.invoke_with_connection_json(
            ["app", "version", "create", "v1", "--force", "--skip-git-check"],
            env=TEST_ENV,
        )

        try:
            # package exist
            package_name = f"{project_name}_pkg_{USER_NAME}".upper()
            app_name = f"{project_name}_{USER_NAME}".upper()
            # app package contains version v1
            expect = snowflake_session.execute_string(
                f"show versions in application package {package_name}"
            )
            actual = runner.invoke_with_connection_json(
                ["app", "version", "list"], env=TEST_ENV
            )
            assert actual.json == row_from_snowflake_session(expect)

            runner.invoke_with_connection_json(
                ["app", "run", "--version", "v1", "--force"], env=TEST_ENV
            )

            expect = row_from_snowflake_session(
                snowflake_session.execute_string(f"desc application {app_name}")
            )
            assert contains_row_with(expect, {"property": "name", "value": app_name})
            assert contains_row_with(expect, {"property": "version", "value": "V1"})
            assert contains_row_with(expect, {"property": "patch", "value": "0"})

            runner.invoke_with_connection_json(
                ["app", "version", "drop", "v1", "--force"],
                env=TEST_ENV,
            )

            # make sure we always delete the package
            result = runner.invoke_with_connection_json(
                ["app", "teardown"],
                env=TEST_ENV,
            )
            assert result.exit_code == 0

        finally:
            # teardown is idempotent, so we can execute it again with no ill effects
            result = runner.invoke_with_connection_json(
                ["app", "teardown", "--force"],
                env=TEST_ENV,
            )
            assert result.exit_code == 0


# Make sure we can create 3+ patches on the same version
@pytest.mark.integration
@enable_definition_v2_feature_flag
# @pytest.mark.parametrize("definition_version", ["v1", "v2"])
@pytest.mark.parametrize("project_definition_files", ["integration"], indirect=True)
def test_nativeapp_version_create_3_patches(
    runner,
    snowflake_session,
    project_definition_files: List[Path],
):
    project_name = "integration"
    project_dir = project_definition_files[0].parent
    with pushd(project_dir):
        try:
            package_name = f"{project_name}_pkg_{USER_NAME}".upper()

            # create three patches (deploys too)
            for _ in range(3):
                result = runner.invoke_with_connection_json(
                    ["app", "version", "create", "v1", "--force", "--skip-git-check"],
                    env=TEST_ENV,
                )
                assert result.exit_code == 0

            # app package contains 3 patches for version v1
            expect = row_from_snowflake_session(
                snowflake_session.execute_string(
                    f"show versions in application package {package_name}"
                )
            )
            assert contains_row_with(expect, {"version": "V1", "patch": 0})
            assert contains_row_with(expect, {"version": "V1", "patch": 1})
            assert contains_row_with(expect, {"version": "V1", "patch": 2})

            # drop the version
            result_drop = runner.invoke_with_connection_json(
                ["app", "version", "drop", "v1", "--force"],
                env=TEST_ENV,
            )
            assert result_drop.exit_code == 0

            # ensure there are no versions now
            actual = runner.invoke_with_connection_json(
                ["app", "version", "list"], env=TEST_ENV
            )
            assert len(actual.json) == 0

            # make sure we always delete the package
            result = runner.invoke_with_connection_json(
                ["app", "teardown"],
                env=TEST_ENV,
            )
            assert result.exit_code == 0

            expect = snowflake_session.execute_string(
                f"show application packages like '{package_name}'"
            )
            assert not_contains_row_with(
                row_from_snowflake_session(expect), {"name": package_name}
            )

        finally:
            # teardown is idempotent, so we can execute it again with no ill effects
            result = runner.invoke_with_connection_json(
                ["app", "teardown", "--force"],
                env=TEST_ENV,
            )
            assert result.exit_code == 0


@pytest.mark.integration
@enable_definition_v2_feature_flag
@pytest.mark.parametrize(
    "project_definition_files", ["integration", "integration_v2"], indirect=True
)
def test_nativeapp_version_create_patch_is_integer(
    runner,
    snowflake_session,
    project_definition_files: List[Path],
):
    project_name = "integration"
    project_dir = project_definition_files[0].parent
    with pushd(project_dir):
        try:
            package_name = f"{project_name}_pkg_{USER_NAME}".upper()

            # create initial version
            result = runner.invoke_with_connection_json(
                ["app", "version", "create", "v1", "--force", "--skip-git-check"],
                env=TEST_ENV,
            )
            assert result.exit_code == 0

            # create non-integer patch
            result = runner.invoke_with_connection_json(
                [
                    "app",
                    "version",
                    "create",
                    "v1",
                    "--force",
                    "--skip-git-check",
                    "--patch",
                    "foo",
                ],
                env=TEST_ENV,
            )
            assert result.exit_code == 2
            assert (
                "Invalid value for '--patch': 'foo' is not a valid integer."
                in result.output
            )

            # create integer patch
            result = runner.invoke_with_connection_json(
                [
                    "app",
                    "version",
                    "create",
                    "v1",
                    "--force",
                    "--skip-git-check",
                    "--patch",
                    "1",
                ],
                env=TEST_ENV,
            )
            assert result.exit_code == 0

            # drop the version
            result_drop = runner.invoke_with_connection_json(
                ["app", "version", "drop", "v1", "--force"],
                env=TEST_ENV,
            )
            assert result_drop.exit_code == 0

            # ensure there are no versions now
            actual = runner.invoke_with_connection_json(
                ["app", "version", "list"], env=TEST_ENV
            )
            assert len(actual.json) == 0

            # make sure we always delete the package
            result = runner.invoke_with_connection_json(
                ["app", "teardown"],
                env=TEST_ENV,
            )
            assert result.exit_code == 0

            expect = snowflake_session.execute_string(
                f"show application packages like '{package_name}'"
            )
            assert not_contains_row_with(
                row_from_snowflake_session(expect), {"name": package_name}
            )
        finally:
            # teardown is idempotent, so we can execute it again with no ill effects
            result = runner.invoke_with_connection_json(
                ["app", "teardown", "--force"],
                env=TEST_ENV,
            )
            assert result.exit_code == 0


# Tests creating a version for a package that was not created by the CLI
# (doesn't have the magic CLI comment)
@pytest.mark.integration
@enable_definition_v2_feature_flag
@pytest.mark.parametrize(
    "project_definition_files", ["integration", "integration_v2"], indirect=True
)
def test_nativeapp_version_create_package_no_magic_comment(
    runner,
    snowflake_session,
    snapshot,
    project_definition_files: List[Path],
):
    project_name = "integration"
    project_dir = project_definition_files[0].parent
    with pushd(project_dir):
        result_create_abort = runner.invoke_with_connection_json(
            ["app", "deploy"],
            env=TEST_ENV,
        )
        assert result_create_abort.exit_code == 0

        try:
            # package exist
            package_name = f"{project_name}_pkg_{USER_NAME}".upper()
            assert contains_row_with(
                row_from_snowflake_session(
                    snowflake_session.execute_string(
                        f"show application packages like '{package_name}'",
                    )
                ),
                dict(name=package_name),
            )

            assert contains_row_with(
                row_from_snowflake_session(
                    snowflake_session.execute_string(
                        f"alter application package {package_name} set comment = 'not the magic comment'"
                    )
                ),
                dict(status="Statement executed successfully."),
            )

            # say no
            result_create_abort = runner.invoke_with_connection(
                ["app", "version", "create", "v1", "--skip-git-check", "--interactive"],
                env=TEST_ENV,
                input="n\n",
            )
            assert result_create_abort.exit_code == 1
            assert (
                f"An Application Package {package_name} already exists in account "
                "that may have been created without Snowflake CLI.".upper()
                in result_create_abort.output.upper()
            )
            assert "Aborted." in result_create_abort.output

            # say yes
            result_create_yes = runner.invoke_with_connection(
                ["app", "version", "create", "v1", "--skip-git-check", "--interactive"],
                env=TEST_ENV,
                input="y\n",
            )
            assert result_create_yes.exit_code == 0
            assert (
                f"An Application Package {package_name} already exists in account "
                "that may have been created without Snowflake CLI.".upper()
                in result_create_yes.output.upper()
            )

            # force
            result_create_force = runner.invoke_with_connection(
                ["app", "version", "create", "v1", "--force", "--skip-git-check"],
                env=TEST_ENV,
            )
            assert result_create_force.exit_code == 0
            assert (
                f"An Application Package {package_name} already exists in account "
                "that may have been created without Snowflake CLI.".upper()
                in result_create_force.output.upper()
            )

            # app package contains version v1 with 2 patches
            actual = runner.invoke_with_connection_json(
                ["app", "version", "list"], env=TEST_ENV
            )
            for row in actual.json:
                # Remove date field
                row.pop("created_on", None)
            assert actual.json == snapshot
        finally:
            # teardown is idempotent, so we can execute it again with no ill effects
            result = runner.invoke_with_connection_json(
                ["app", "teardown", "--force"],
                env=TEST_ENV,
            )
            assert result.exit_code == 0
