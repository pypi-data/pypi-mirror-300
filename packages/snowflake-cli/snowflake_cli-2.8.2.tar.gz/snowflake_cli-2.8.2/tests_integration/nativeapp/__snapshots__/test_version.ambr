# serializer version: 1
# name: test_nativeapp_version_create_package_no_magic_comment[integration]
  list([
    dict({
      'comment': 'Default version used for development. Override for actual deployment.',
      'dropped_on': None,
      'label': 'Dev Version',
      'log_level': 'INFO',
      'patch': 0,
      'review_status': 'NOT_REVIEWED',
      'state': 'READY',
      'trace_level': 'ALWAYS',
      'version': 'V1',
    }),
    dict({
      'comment': 'Default version used for development. Override for actual deployment.',
      'dropped_on': None,
      'label': 'Dev Version',
      'log_level': 'INFO',
      'patch': 1,
      'review_status': 'NOT_REVIEWED',
      'state': 'READY',
      'trace_level': 'ALWAYS',
      'version': 'V1',
    }),
  ])
# ---
# name: test_nativeapp_version_create_package_no_magic_comment[integration_v2]
  list([
    dict({
      'comment': 'Default version used for development. Override for actual deployment.',
      'dropped_on': None,
      'label': 'Dev Version',
      'log_level': 'INFO',
      'patch': 0,
      'review_status': 'NOT_REVIEWED',
      'state': 'READY',
      'trace_level': 'ALWAYS',
      'version': 'V1',
    }),
    dict({
      'comment': 'Default version used for development. Override for actual deployment.',
      'dropped_on': None,
      'label': 'Dev Version',
      'log_level': 'INFO',
      'patch': 1,
      'review_status': 'NOT_REVIEWED',
      'state': 'READY',
      'trace_level': 'ALWAYS',
      'version': 'V1',
    }),
  ])
# ---
